datatree/tests/test_datatree.py

```diff
@@ -175,7 +206,86 @@ def test_getitem_dict_like_selection_access_to_dataset(self):
 
 
 class TestUpdate:
-    ...
+    def test_update_new_named_dataarray(self):
+        da = xr.DataArray(name="temp", data=[0, 50])
+        folder1 = DataTree(name="folder1")
+        folder1.update({"results": da})
+        expected = da.rename("results")
+        xrt.assert_equal(folder1["results"], expected)
+
+
+class TestCopy:
+    def test_copy(self):
+        dt = create_test_datatree()
+
+        for node in dt.root.subtree:
+            node.attrs["Test"] = [1, 2, 3]
+
+        for copied in [dt.copy(deep=False), copy(dt)]:
+            dtt.assert_identical(dt, copied)
+
+            for node, copied_node in zip(dt.root.subtree, copied.root.subtree):
+
+                assert node.encoding == copied_node.encoding
+                # Note: IndexVariable objects with string dtype are always
+                # copied because of xarray.core.util.safe_cast_to_index.
+                # Limiting the test to data variables.
+                # TODO use .data_vars once that property is available
+                data_vars = [v for v in node.variables if v not in node._coord_names]
+                for k in data_vars:
+                    v0 = node.variables[k]
+                    v1 = copied_node.variables[k]
+                    assert source_ndarray(v0.data) is source_ndarray(v1.data)
+                copied_node["foo"] = xr.DataArray(data=np.arange(5), dims="z")
+                assert "foo" not in node
+
+                copied_node.attrs["foo"] = "bar"
+                assert "foo" not in node.attrs
+                assert node.attrs["Test"] is copied_node.attrs["Test"]
+
+    def test_deepcopy(self):
+        dt = create_test_datatree()
+
+        for node in dt.root.subtree:
+            node.attrs["Test"] = [1, 2, 3]
+
+        for copied in [dt.copy(deep=True), deepcopy(dt)]:
+            dtt.assert_identical(dt, copied)
+
+            for node, copied_node in zip(dt.root.subtree, copied.root.subtree):
+                assert node.encoding == copied_node.encoding
+                # Note: IndexVariable objects with string dtype are always
+                # copied because of xarray.core.util.safe_cast_to_index.
+                # Limiting the test to data variables.
+                # TODO use .data_vars once that property is available
+                data_vars = [v for v in node.variables if v not in node._coord_names]
+                for k in data_vars:
+                    v0 = node.variables[k]
+                    v1 = copied_node.variables[k]
+                    assert source_ndarray(v0.data) is source_ndarray(v1.data)
```

Comment:
There's a bug with deepcopying data in the tree here, causing an unhelpful assertion error:

```python
>                   assert source_ndarray(v0.data) is source_ndarray(v1.data)
E                   assert array([6, 7, 8]) is array([6, 7, 8])
E                    +  where array([6, 7, 8]) = source_ndarray(array([6, 7, 8]))
E                    +    where array([6, 7, 8]) = <xarray.Variable (y: 3)>\narray([6, 7, 8]).data
E                    +  and   array([6, 7, 8]) = source_ndarray(array([6, 7, 8]))
E                    +    where array([6, 7, 8]) = <xarray.Variable (y: 3)>\narray([6, 7, 8]).data

datatree/tests/test_datatree.py:265: AssertionError
```

Reply:
Aren't you checking here that v1 has been copied and therefore is not referring to the same object as v0? If that's the case, I think you need:

```python
assert source_ndarray(v0.data) is not source_ndarray(v1.data)
```
(which passes the test).

Same on line 271.

Reply:
Aren't you checking here that v1 has been copied and therefore is not referring to the same object as v0? If that's the case, I think you need:

```python
assert source_ndarray(v0.data) is not source_ndarray(v1.data)
```
(which passes the test).

Same on line 271.

Reply:
Good spot, thank you!